pandas/tests/indexes/timedeltas/test_setops.py

```diff
@@ -73,3 +76,92 @@ def test_intersection_bug_1708(self):
+        # GH 24471 test for non overlap the intersection should be zero length
+        index_1 = timedelta_range('1 day', periods=period_1, freq='h')
+        index_2 = timedelta_range('1 day', periods=period_2, freq='h')
+        inter = index_1.intersection(index_2, sort=sort)
+        tm.assert_index_equal(timedelta_range('1 day', periods=0, freq='h'),
+                              inter)
+
+    @pytest.mark.parametrize('sort', [None, False])
+    def test_zero_length_input_index(self, sort):
+        index_1 = timedelta_range('1 day', periods=0, freq='h')
+        index_2 = timedelta_range('1 day', periods=3, freq='h')
+        inter = index_1.intersection(index_2, sort=sort)
+        assert index_1 is not inter
+        assert index_2 is not inter
+
+    @pytest.mark.parametrize(
+        "rng, expected",
+        # if target has the same name, it is preserved
+        [
+            (timedelta_range('1 day', periods=5, freq='h', name='idx'),
+             timedelta_range('1 day', periods=4, freq='h', name='idx')),
+            # if target name is different, it will be reset
+            (timedelta_range('1 day', periods=5, freq='h', name='other'),
+             timedelta_range('1 day', periods=4, freq='h', name=None)),
+            # if no overlap exists return empty index
+            (timedelta_range('1 day', periods=10, freq='h', name='idx')[5:],
+             TimedeltaIndex([], name='idx'))])
+    @pytest.mark.parametrize("sort", [None, False])
+    def test_intersection(self, rng, expected, sort):
+        # GH 4690 (with tz)
+        base = timedelta_range('1 day', periods=4, freq='h', name='idx')
+        result = base.intersection(rng, sort=sort)
+        if sort is None:
+            expected = expected.sort_values()
+        tm.assert_index_equal(result, expected)
+        assert result.name == expected.name
+        assert result.freq == expected.freq
+
+    @pytest.mark.parametrize(
+        "rng, expected",
+        # part intersection works
+        [
+            (TimedeltaIndex(['5 hour', '2 hour', '4 hour', '9 hour'],
+                            name='idx'),
+             TimedeltaIndex(['2 hour', '4 hour'], name='idx')),
+            # reordered part intersection
+            (TimedeltaIndex(['2 hour', '5 hour', '5 hour', '1 hour'],
+                            name='other'),
+             TimedeltaIndex(['1 hour', '2 hour'], name=None)),
+            # reveresed index
+            (TimedeltaIndex(['1 hour', '2 hour', '4 hour', '3 hour'],
+                            name='idx')[::-1],
+             TimedeltaIndex(['1 hour', '2 hour', '4 hour', '3 hour'],
+                            name='idx'))])
+    @pytest.mark.parametrize("sort", [None, False])
+    def test_intersection_non_monotonic(self, rng, expected, sort):
+        # 24471 non-monotonic
+        base = TimedeltaIndex(['1 hour', '2 hour', '4 hour', '3 hour'],
+                              name='idx')
+        result = base.intersection(rng, sort=sort)
```

Comment:
need a test to assert that a 0-len self or other is actually a copy.

Reply:
@jreback While digging into this test a bit more I came across the following case. It turns out that we don't actually hit the 0-length intersection case because 

https://github.com/pandas-dev/pandas/blob/ed6c61ddd15d4256a692ce63c2d52bd6db0d0b4f/pandas/core/indexes/datetimelike.py#L550-L558

the `not other.freq.isAnchored()` will always trigger this elif argument for timedeltas that have a tick based frequency. As far as I can tell this is because frequencies that inherent from the ``Tick`` class will by default return False for the `isAnchored()` function. I am assuming here that isAnchored refers to a timeindex etc to be anchored to a specific date/weekday? The documentation here is not quiet clear.

https://github.com/pandas-dev/pandas/blob/ed6c61ddd15d4256a692ce63c2d52bd6db0d0b4f/pandas/tseries/offsets.py#L307-L310

and for Tick class specifically:

https://github.com/pandas-dev/pandas/blob/ed6c61ddd15d4256a692ce63c2d52bd6db0d0b4f/pandas/tseries/offsets.py#L2341-L2342

Maybe you or @reidy-p have an idea why we testing for anchoring here or shine some more light on this function/property. 
There are currently three possibilities I see this moving forward:
1)Move the checks for 0-length index before the elif that includes the `isAnchored()` check so that we make sure this is hit.
2) Maybe we don't need the check for `isAnchored()` before and can move it further back.
3) Design the test such that `isAnchored()` is not triggered.

Some input would be welcome.

Reply:
so I would try to move the checks right below the .equals() check and see what happens

Reply:
add, also just wanted to thank you for your patience and continuous comments!